core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ClusterService.java

```diff
@@ -948,28 +950,41 @@ public Cluster updateCreationDateOnCluster(Cluster cluster) {
-            throw new TransactionRuntimeExecutionException(e);
+        if (!connector.clusterStatusService().isCMRunning()) {
+            InstanceMetaData cmInstance = notTerminatedInstanceMetaDatas.stream().filter(InstanceMetaData::getClusterManagerServer).findFirst()
+                    .orElseThrow(() -> new CloudbreakServiceException("Cluster manager inaccessible, and the corresponding instance metadata not found."));
+            cmInstance.setInstanceStatus(InstanceStatus.SERVICES_UNHEALTHY);
+            cmInstance.setStatusReason("Cluster manager inaccessible.");
+            instanceMetaDataService.save(cmInstance);
+            eventService.fireCloudbreakEvent(stack.getId(), AVAILABLE.name(), cloudbreakMessagesService.getMessage(Msg.CLUSTER_HOST_STATUS_UPDATED.code(),
+                    Arrays.asList(cmInstance.getDiscoveryFQDN(), cmInstance.getInstanceStatus().name())));
+            return stack.getCluster();
+        } else {
+            Map<String, ClusterManagerState> hostStatuses = connector.clusterStatusService().getExtendedHostStatuses();
+            try {
+                return transactionService.required(() -> {
+                    List<InstanceMetaData> updatedInstanceMetaData = notTerminatedInstanceMetaDatas.stream()
+                            .filter(instanceMetaData -> InstanceStatus.SERVICES_RUNNING.equals(instanceMetaData.getInstanceStatus()))
+                            .map(instanceMetaData -> {
+                                ClusterManagerState clusterManagerState = hostStatuses.get(instanceMetaData.getDiscoveryFQDN());
+                                InstanceStatus newState = ClusterManagerStatus.HEALTHY.equals(clusterManagerState.getClusterManagerStatus()) ?
+                                        InstanceStatus.SERVICES_HEALTHY : InstanceStatus.SERVICES_UNHEALTHY;
+                                instanceMetaData.setInstanceStatus(newState);
+                                instanceMetaData.setStatusReason(clusterManagerState.getStatusReason());
+                                return instanceMetaData;
+                            }).collect(Collectors.toList());
+                    instanceMetaDataService.saveAll(updatedInstanceMetaData);
+                    String hostNotificationMessage = updatedInstanceMetaData.stream()
+                            .map(instanceMetaData -> cloudbreakMessagesService.getMessage(Msg.CLUSTER_HOST_STATUS_UPDATED.code(),
+                                    Arrays.asList(instanceMetaData.getDiscoveryFQDN(), instanceMetaData.getInstanceStatus().name())))
+                            .collect(joining("; "));
+                    eventService.fireCloudbreakEvent(stack.getId(), AVAILABLE.name(), hostNotificationMessage);
+                    return stack.getCluster();
+                });
+            } catch (TransactionExecutionException e) {
+                throw new TransactionRuntimeExecutionException(e);
+            }
```

Comment:
this method got a little bit big and hard to overview, please refactor it into some smaller separate method